### PR TITLE
Discussion around journaling

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

Discuss the different invariants and the assumptions for journaling to work.

## Proposal

TBD

### CODE REVIEW

File: linera-views/src/journaling.rs [hunk 1]
```diff
@@ -231,28 +250,62 @@ where
...
                     (true, true)
                 } else {
                     let next_block_size =
                         block_size + iter.next_value_size()? + block_batch.overhead_size();
                     let next_transaction_size = transaction_size + next_block_size;
                     let transaction_flush = next_transaction_size > K::MAX_BATCH_TOTAL_SIZE;
-                    let block_flush = transaction_flush || next_block_size > K::MAX_VALUE_SIZE;
+                    let block_flush = transaction_flush
+                        || block_batch.len() == K::MAX_BATCH_SIZE - 2
+                        || next_block_size > K::MAX_VALUE_SIZE;
```
Comment (ma2bd) [inline linera-views/src/journaling.rs:308]: I'm inclined to blame this existing inequality (rather than my change line 299) for the failed test.

**Comment by afck** [comment]
I still think this here might be an issue: https://github.com/linera-io/linera-protocol/pull/1483#discussion_r1451414720
Not sure if it's related.

**Comment by ma2bd** [comment]
> I still think this here might be an issue: [#1483 (comment)](https://github.com/linera-io/linera-protocol/pull/1483#discussion_r1451414720) Not sure if it's related.

Let me add a fix for this one here.

File: linera-views/src/journaling.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-views/src/journaling.rs:308]: I am also more inclined to think the correction is needed, but I will do more extensive checks.

File: linera-views/src/journaling.rs [hunk 2]
```diff
@@ -200,7 +200,26 @@ where
...
+    /// (1) each block contains at most `K::MAX_BATCH_SIZE - 2` operations;
+    ///
+    /// (2) the total size of the all operations in a block doesn't exceed:
+    ///     `K::MAX_BATCH_TOTAL_SIZE - sizeof(block_key) - sizeof(header_key) - sizeof(bcs_header)`
+    ///
+    /// (3) every operation in a block satisfies the contraints on individual database operations
+    ///     represented by `K::MAX_KEY_SIZE` and `K::MAX_VALUE_SIZE`.
+    ///
+    /// (4) `block_key` and `header_key` don't exceed `K::MAX_KEY_SIZE` and `bcs_header`
+    /// doesn't exceed `K::MAX_VALUE_SIZE`.
```
Comment (afck) [inline linera-views/src/journaling.rs:222]:
```suggestion
    ///     doesn't exceed `K::MAX_VALUE_SIZE`.
```

(Or the other way round. Indentation is also a bit inconsistent in the other doc comments.)

**Comment by ma2bd** [comment]
The error in CI is 600 bytes difference. It could be the JSON encoding overhead for what I know.

I created the non-urgent task #1496 btw

**Comment by ma2bd** [comment]
I'm moving the first commit of this PR to #1502

File: linera-views/src/journaling.rs [hunk 3]
```diff
@@ -231,28 +250,77 @@ where
...
         base_key: &[u8],
     ) -> Result<JournalHeader, K::Error> {
-        let mut iter = simplified_batch.into_iter();
+        let header_key = get_journaling_key(base_key, KeyTag::Journal as u8, 0)?;
+        let header_key_len = header_key.len();
+        let header_len = bcs::serialized_size(&JournalHeader::default())?;
+        // Currently the same length.
+        let block_key_len = header_key.len();
+        // Each block in a transaction comes with a key.
+        let max_transaction_size = K::MAX_BATCH_TOTAL_SIZE - K::MAX_BATCH_SIZE * block_key_len;
```
Comment (MathieuDutSik) [inline linera-views/src/journaling.rs:298]: I actually can get an overflow here for `ScyllaDb`.
The thing is that for `ScyllaDb` the size of the batch is limited to 16M, but the number of transactions inside is not upper bound. So `MAX_BATCH_TOTAL_SIZE=16*1024*1024` and `MAX_BATCH_SIZE = USIZE::MAX`.

**Pull request closed without merging**